File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = 'aeiouAEIOU'

        slist = list(s)

        i, j = 0, len(s) - 1

        while i < j:
            while i < j and slist[i] not in vowels:
                i += 1
            while j > i and slist[j] not in vowels:
                j -= 1
            
            slist[i], slist[j] = slist[j], slist[i]

            i += 1
            j -= 1
        
        return ''.join(slist)

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.js
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function(s) {
    
    let i = 0;
    let j = s.length - 1;

    const res = Array.from(s);

    const set = new Set(['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']);

    while (i < j) {

        while (i < j && !set.has(res[i])) i++;
        while (i < j && !set.has(res[j])) j--;
        const temp = res[i];
        res[i] = res[j];
        res[j] = temp;
        i++;
        j--;
    }

    return res.join('');
};